src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/QueryGenerators/SqlQueryGenerator.cs

```diff
@@ -424,8 +436,65 @@ public object VisitTable(TableExpression tableExpression, SearchOptions context)
+                        var reourceId = Model.GetResourceTypeId(includeExpression.ResourceType);
+                        if (includeExpression.Reversed && includeExpression.Iterate)
+                        {
+                            if (includeExpression.TargetResourceType != null)
+                            {
+                                reourceId = Model.GetResourceTypeId(includeExpression.TargetResourceType);
+                            }
+                            else if (includeExpression.ReferenceSearchParameter?.TargetResourceTypes?.Count > 0)
+                            {
+                                reourceId = Model.GetResourceTypeId(includeExpression.ReferenceSearchParameter.TargetResourceTypes.ToList().First());
+                            }
+                        }
+
                         delimited.BeginDelimitedElement().Append(VLatest.ReferenceSearchParam.ResourceTypeId, table)
-                            .Append(" = ").Append(Parameters.AddParameter(VLatest.ReferenceSearchParam.ResourceTypeId, Model.GetResourceTypeId(includeExpression.ResourceType)));
+                        .Append(" = ").Append(Parameters.AddParameter(VLatest.ReferenceSearchParam.ResourceTypeId, reourceId));
+
+                        // Get FROM ctes
+                        string fromCte = _cteMainSelect;
+                        if (includeExpression.Iterate)
+                        {
+                            // Include Iterate
+                            if (!includeExpression.Reversed)
+                            {
+                                // _include:iterate may appear without a preceding _include, in case of circular reference
+                                // On that case, the fromCte is _cteMainSelect
+                                if (TryGetIncludeCtes(includeExpression.SourceResourceType, out _includeFromCtes))
+                                {
+                                    fromCte = _includeFromCtes[++_curFromCteIndex];
+                                }
+                            }
+
+                            // RevInclude Iterate
+                            else
+                            {
+                                if (includeExpression.TargetResourceType != null)
+                                {
+                                    if (TryGetIncludeCtes(includeExpression.TargetResourceType, out _includeFromCtes))
+                                    {
+                                        fromCte = _includeFromCtes[++_curFromCteIndex];
+                                    }
+                                }
+                                else if (includeExpression.ReferenceSearchParameter?.TargetResourceTypes != null)
+                                {
+                                    // Assumes TargetResourceTypes is of length 1. Otherwise, a BadRequest would have been thrown earlier for _revinclude:iterate
```

Comment:
I would also throw here in any case, in the future someone might remove the other exception throw without knowing about this dependency.

Reply:
No need to, that's why we have a test for it :)

If someone removes it in SearchOpetionsFactory, the test will fail :)